### PR TITLE
[KIECLOUD-677] Create a new Jenkins seed job for operators

### PR DESCRIPTION
**Pull request by davidesalerno**
Signed-off-by: Davide Salerno <dsalerno@redhat.com>

**Thank you for submitting this pull request**

**JIRA**: [link](https://issues.redhat.com/browse/KIECLOUD-677)

<pre>
How to retest a PR or trigger a specific build:

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
</pre>

### CODE REVIEW

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+            stringParam("IMAGE_REPO", "${imageRepo}")
+            stringParam("IMAGE_BRANCH", "${imageBranch}")
+            stringParam("IMAGE_SUBDIR", "${imageSubdir}")
+            stringParam("GIT_USER", "${gitUser}")
+            stringParam("GIT_EMAIL", "${gitEmail}")
+            stringParam("CEKIT_CACHE_LOCAL", "${cekitCacheLocal}")
+            stringParam("VERBOSE", "${verbose}")
+            stringParam("GITHUB_ORG_UNIT", "${githubOrgUnit}")
+            stringParam("BAMOE_BA_OPERTOR_BRANCH", "${bamoeBAOperatorBranch}")
+            stringParam("BAMOE_KOGITO_OPERTOR_BRANCH", "${bamoeKogitoOperatorBranch}")
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:58]:
```suggestion
            stringParam('BUILD_DATE', "${buildDate}")
            stringParam('PROD_VERSION', "${prodVersion}")
            stringParam('PROD_COMPONENT', "${Component}")
            stringParam('OSBS_BUILD_TARGET', "${osbsBuildTarget}")
            stringParam('CEKIT_BUILD_OPTIONS', "${cekitBuildOptions}")
            stringParam('KERBEROS_PRINCIPAL', "${kerberosPrincipal}")
            stringParam('OSBS_BUILD_USER', "${osbsBuildUser}")
            stringParam('KERBEROS_KEYTAB', "${kerberosKeytab}")
            stringParam('KERBEROS_CRED', "${kerberosCred}")
            stringParam('IMAGE_REPO', "${imageRepo}")
            stringParam('IMAGE_BRANCH', "${imageBranch}")
            stringParam('IMAGE_SUBDIR', "${imageSubdir}")
            stringParam('GIT_USER', "${gitUser}")
            stringParam('GIT_EMAIL', "${gitEmail}")
            stringParam('CEKIT_CACHE_LOCAL', "${cekitCacheLocal}")
            stringParam('VERBOSE', "${verbose}")
            stringParam('GITHUB_ORG_UNIT', "${githubOrgUnit}")
            stringParam('BAMOE_BA_OPERTOR_BRANCH', "${bamoeBAOperatorBranch}")
            stringParam('BAMOE_KOGITO_OPERTOR_BRANCH', "${bamoeKogitoOperatorBranch}")
```

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                          // The build script is in the image, but build.sh and build-overrides.sh which it calls will be downloaded
+                          def build_command = "build-osbs.sh"
+                         
+                          // Create the build command
+                          validateParameters(REQUIRED_BUILD_PARAMETERS, OPTIONAL_BUILD_PARAMETERS)
+
+                          for(param in REQUIRED_BUILD_PARAMETERS){
+                              def arg = param.value[0]
+                              def flag = param.value[1]
+                              build_command += " ${flag} ${arg}"
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:140]:
```suggestion
                              build_command += " ${param.value[1]} ${param.value[0]}"
```

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+                          }
+
+                          for(param in OPTIONAL_BUILD_PARAMETERS){
+                              def arg = param.value[0]
+                              def flag = param.value[1]
+                              if(arg) build_command+= " ${flag} ${arg}"
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:146]:
```suggestion
                              build_command+= arg ? " ${flag} ${arg}" : ''
```

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+                          }
+
+                          for(param in OPTIONAL_BUILD_SWITCHES){
+                              def arg = param.value[0]
+                              def flag = param.value[1]
+                              if(arg == "true") build_command+= " ${flag}"
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:152]:
```suggestion
                              build_command+= arg == 'true' ? " ${flag}" : ''
```

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+                          build_command +=" -w ${WORKSPACE}"
+                          
+                          operator_repo_name = getOperatorRepoName(PROD_COMPONENT)
+                          
+                          operator_branch = getOperatorBranch(PROD_COMPONENT)
+                          
+                          // Checking out from git the Operator repo
+                          sh "git config --global user.email $GIT_EMAIL"
+                          sh "git config --global user.name $GIT_AUTHOR"
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:163]:
```suggestion
                          sh 'git config --global user.email $GIT_EMAIL'
                          sh 'git config --global user.name $GIT_AUTHOR'
```

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+                          githubscm.getRepositoryScm(operator_repo_name, GITHUB_ORG_UNIT, operator_branch)
+                          
+                          // Run the build script that should be into the operator hack folder
+                          dir("hack") {
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:167]:
```suggestion
                          dir('hack') {
```

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+                            sh "source ~/virtenvs/cekit/bin/activate && $build_command"
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:168]:
```suggestion
                            sh "source ~/virtenvs/cekit/bin/activate && ${build_command}"
```

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+                          }
+                          // post processing
+                      }
+
+                      node("osbs-builder"){
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:173]:
```suggestion
                      node('osbs-builder'){
```

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+                          stage("building ${PROD_COMPONENT}") {
+                              ws {
+                                  // Add the working directory to the current path so any scripts dropped there are on the path
+                                  withEnv(["PATH+W=$WORKSPACE"]) {
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:177]:
```suggestion
                                  withEnv(['PATH+W=$WORKSPACE']) {
```
is this what you mean, or `withEnv(["PATH+W=${WORKSPACE}"]) {`

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+                                      sh 'rm -rf ${WORKSPACE}/{*,.*} || true'
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:178]:
```suggestion
                                      sh "rm -rf ${WORKSPACE}/{*,.*} || true"
```
you replace from groovy to replace ` ${WORKSPACE}` right?

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+                                      if (env.KERBEROS_CRED) {
+                                          withCredentials([usernamePassword(credentialsId: env.KERBEROS_CRED, usernameVariable: 'USERNAME', passwordVariable: 'PASSWORD')]) {
+                                              mainProcess(USERNAME, PASSWORD, "")
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:181]:
```suggestion
                                              mainProcess(USERNAME, PASSWORD, '')
```

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+                                          }
+                                      } else if (env.KERBEROS_KEYTAB) {
+                                          withCredentials([file(credentialsId: env.KERBEROS_KEYTAB, variable: 'FILE')]) {
+                                              if (!env.KERBEROS_PRINCIPAL) {
+                                                    echo "Reading the Kerberos Principal from provided Keytab..."
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:186]:
```suggestion
                                                    echo 'Reading the Kerberos Principal from provided Keytab...'
```

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+                                                    def get_principal_from_file = sh(returnStdout: true, script: \'\'\'
+                                                        #!/bin/bash
+                                                        klist -kt $FILE |grep REDHAT.COM | awk -F" " \'NR==1{print $4}\'
+                                                    \'\'\')
+                                                    env.KERBEROS_PRINCIPAL = get_principal_from_file.trim()
+                                              }  
+                                              mainProcess(env.KERBEROS_PRINCIPAL, "", FILE)
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:193]:
```suggestion
                                              mainProcess(env.KERBEROS_PRINCIPAL, '', FILE)
```

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+                                          }
+                                      } else {
+                                          error "Either KERBEROS_PRINCIPAL and KERBEROS_KEYTAB must be specified, or user/password with KERBEROS_CRED"
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:196]:
```suggestion
                                          error 'Either KERBEROS_PRINCIPAL and KERBEROS_KEYTAB must be specified, or user/password with KERBEROS_CRED'
```

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 15]
```diff
@@ -0,0 +1,207 @@
+package osbs
+
+import org.kie.jenkins.jobdsl.Constants
+
+folder("OSBS")
+def folderPath = "OSBS"
+// Job Description
+String jobDescription = "Job responsible for seed jobs to building rhpam and bamoe ba and kogito operator images"
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:8]:
```suggestion
def folderPath = 'OSBS'
folder(folderPath)
// Job Description
String jobDescription = 'Job responsible for seed jobs to building rhpam and bamoe ba and kogito operator images'
```

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+                                  "CEKIT_CACHE_LOCAL": [CEKIT_CACHE_LOCAL, '-l'],
+                          ]
+
+                          def OPTIONAL_BUILD_SWITCHES = [
+                                  "VERBOSE": [VERBOSE, '-g'],
+                          ]
+
+       
+                          // The build script is in the image, but build.sh and build-overrides.sh which it calls will be downloaded
+                          def build_command = "build-osbs.sh"
```
Comment (Ginxo) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:132]:
```suggestion
                          def REQUIRED_BUILD_PARAMETERS = [
                                  'PROD_VERSION': [PROD_VERSION, '-v'],
                                  'PROD_COMPONENT': [PROD_COMPONENT, '-c'],
                                  'OSBS_BUILD_TARGET': [OSBS_BUILD_TARGET, '-t'],
                          ]

                          def OPTIONAL_BUILD_PARAMETERS = [
                                  'KERBEROS_PASSWORD': [password, '-s'],
                                  'KERBEROS_PRINCIPAL': [user, '-p'],
                                  'KERBEROS_KEYTAB': [keytab, '-k'],
                                  'OSBS_BUILD_USER': [OSBS_BUILD_USER, '-i'],
                                  'BUILD_DATE': [BUILD_DATE, '-b'],
                                  'GIT_USER': [GIT_USER, '-u'],
                                  'GIT_EMAIL': [GIT_EMAIL, '-e'],
                                  'CEKIT_BUILD_OPTIONS': [CEKIT_BUILD_OPTIONS, '-o'],
                                  'CEKIT_CACHE_LOCAL': [CEKIT_CACHE_LOCAL, '-l'],
                          ]

                          def OPTIONAL_BUILD_SWITCHES = [
                                  'VERBOSE': [VERBOSE, '-g'],
                          ]

       
                          // The build script is in the image, but build.sh and build-overrides.sh which it calls will be downloaded
                          def build_command = 'build-osbs.sh'
```

**Review by Ginxo — CHANGES_REQUESTED** [review]
few technical suggestions, please check

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+                            default:
+                               error "${prodComponent} not supported."
+                          }
+                      }
+                      
+                      // Function to retrieve from the PROD_COMPONENT name the related repo branch
+                      private String getOperatorBranch(prodComponent){
+                          switch(prodComponent){
+                            case { it.startsWith('rhpam') }:
+                                return 'main'
```
Comment (spolti) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:93]: for kogito I guess it is not the main branch, it should be 1.13.x and for bamoe 1.13.x-blue

for ba-operator main stands for the next development cycle, so main branch is correct
and for. bamoe ba, it is release-v7.13.x-blue

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy (hunk 10 above)
Comment (davidesalerno) [reply job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:182]: yes

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy (hunk 17 above)
Comment (davidesalerno) [reply job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:97]: I added the constant RHPAM_KOGITO_OPERTOR_BRANCH for the Kogito Operator branch.

You can see it here https://github.com/kiegroup/kie-ci/pull/1407/files#diff-9d55ce9075a526a3684617ae54e1f15ee81673ba5c89794b916086a8ffacc4d9R69

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy (hunk 9 above)
Comment (davidesalerno) [reply job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy]: I would like to do what we described in the comment and so add the workspace (current directory) to the path

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+                          }
+                      }
+                      
+                      // Function to retrieve from the PROD_COMPONENT name the related repo branch
+                      private String getOperatorBranch(prodComponent){
+                          switch(prodComponent){
+                            case { it.startsWith('rhpam-ba') }:
+                                return 'main'
+                            case { it.startsWith('rhpam-kogito') }:
+                                return RHPAM_KOGITO_OPERTOR_BRANCH
```
Comment (spolti) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:99]: shouldn't these values be rhpamKogitoOperatorBranch or Constants.RHPAM_KOGITO_OPERTOR_BRANCH?
I think the first option would be better, wdyt?

Comment (davidesalerno) [reply job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:99]: I'm defining that variable using the [stringParam](https://jenkinsci.github.io/job-dsl-plugin/#method/javaposse.jobdsl.dsl.helpers.BuildParametersContext.stringParam) function at [line 58](https://github.com/kiegroup/kie-ci/pull/1407/files#diff-02c3e3fd16762232fc76b897b899c1f763fb313735a8159e243aced4da65eb44R58) using the same approach we used for the other variables. Should we change it?

**Review by rgdoliveira — APPROVED** [review]

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy [hunk 19]
```diff
@@ -0,0 +1,207 @@
+package osbs
+
+import org.kie.jenkins.jobdsl.Constants
+
+def folderPath = 'OSBS'
```
Comment (spolti) [inline job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:5]:
```suggestion
def folderPath = 'OSBS/operators'
```

**Review by spolti — APPROVED** [review]

File: job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy (hunk 18 above)
Comment (lampajr) [reply job-dsls/jobs/osbs/operators_osbs_seed_cloud_jobs.groovy:99]: IMHO this is fine as it is, we are using the same approach on all other jobs

**Review by lampajr — APPROVED** [review]

**Pull request merged**